File: lib-androscope/src/main/java/nl/ngti/androscope/server/AndroscopeHttpServer.java
package nl.ngti.androscope.server;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.wifi.WifiInfo;
import android.net.wifi.WifiManager;
import android.text.format.Formatter;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import fi.iki.elonen.NanoHTTPD;
import nl.ngti.androscope.utils.AndroscopeMetadata;

/**
 * Custom HTTP server. Displays the structure of the database.
 */
public class AndroscopeHttpServer extends NanoHTTPD {

    private final Context mContext;

    private final ResponseFactory mResponseFactory;

    private AndroscopeHttpServer(Context context, int httpPort, AndroscopeMetadata metadata) {
        super(httpPort);

        mContext = context.getApplicationContext();

        mResponseFactory = new ResponseFactory(mContext, metadata);
    }

    @NonNull
    public static AndroscopeHttpServer newInstance(Context context) {
        final AndroscopeMetadata metadata = AndroscopeMetadata.fromContext(context);
        final int httpPort = metadata.getHttpPort();
        return new AndroscopeHttpServer(context, httpPort, metadata);
    }

    @Override
    public Response serve(final IHTTPSession session) {
        Response response = createResponse(session);
        if (response == null) {
            response = super.serve(session);
        } else {
            response.addHeader("Access-Control-Allow-Origin", "*");
        }
        return response;
    }

    @NonNull
    public String getIpAddress() {
        @SuppressLint("WifiManagerPotentialLeak") final WifiManager myWifiManager =
                (WifiManager) mContext.getSystemService(Context.WIFI_SERVICE);
        //noinspection ConstantConditions
        final WifiInfo myWifiInfo = myWifiManager.getConnectionInfo();
        final int myIp = myWifiInfo.getIpAddress();
        return Formatter.formatIpAddress(myIp);
    }

    @Nullable
    private Response createResponse(IHTTPSession session) {
        return mResponseFactory.getResponse(session);
    }
}

File: lib-androscope/src/main/java/nl/ngti/androscope/server/AndroscopeHttpServer.kt
package nl.ngti.androscope.server

import android.content.Context
import android.net.wifi.WifiManager
import android.text.format.Formatter
import fi.iki.elonen.NanoHTTPD
import nl.ngti.androscope.BuildConfig
import nl.ngti.androscope.utils.AndroscopeMetadata
import nl.ngti.androscope.utils.AndroscopeMetadata.Companion.fromContext

/**
 * Custom HTTP server. Displays the structure of the database.
 */
internal class AndroscopeHttpServer private constructor(
        private val context: Context,
        httpPort: Int,
        metadata: AndroscopeMetadata
) : NanoHTTPD(httpPort) {

    private val responseFactory = ResponseFactory(context, metadata)

    override fun serve(session: IHTTPSession): Response? {
        return responseFactory.getResponse(session)?.apply {
            if (BuildConfig.DEBUG) {
                // Allows development with Angular running on desktop
                addHeader("Access-Control-Allow-Origin", "*")
            }
        }
    }

    val ipAddress: String
        get() {
            val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
            val wifiInfo = wifiManager.connectionInfo
            val myIp = wifiInfo.ipAddress
            // We don't use IPv6 addresses
            @Suppress("DEPRECATION")
            return Formatter.formatIpAddress(myIp)
        }

    companion object {

        fun newInstance(context: Context): AndroscopeHttpServer {
            val metadata = fromContext(context)
            val httpPort = metadata.httpPort
            return AndroscopeHttpServer(context.applicationContext, httpPort, metadata)
        }
    }
}
